File: src/ru/edu/vtisov/tree/bfs/shortestPathBinaryMatrix.go
// https://leetcode.com/problems/shortest-path-in-binary-matrix/description/

var pos = [][]int{
        []int{-1, -1},
        []int{-1, 0},
        []int{0, -1},
        []int{0, 1},
        []int{1, 0},
        []int{1, -1},
        []int{-1, 1},
        []int{1, 1},
    }

type Point struct {
    x, y, count int
}

func shortestPathBinaryMatrix(grid [][]int) int {
    n := len(grid)
    if grid[0][0] != 0 || grid[n - 1][n - 1] != 0 {
        return -1
    }

    queue := []Point{{ x: 0, y: 0, count: 1 }}

    for len(queue) > 0 {
        cur := queue[0]
        queue = queue[1:]

        if cur.x == n - 1 && cur.y == n - 1 {
            return cur.count
        }
        for _, p := range pos {
            i := cur.x + p[0]
            j := cur.y + p[1]
            if i >= 0 && j >= 0 && i < n && j < n && grid[i][j] != 1 {
                grid[i][j] = 1
                queue = append(queue, Point{x: i, y: j, count: cur.count + 1})
            }
        }
    }
    return -1
}

File: src/ru/edu/vtisov/graph/bfs/ShortestPathBinaryMatrix.java
// https://leetcode.com/problems/shortest-path-in-binary-matrix/

public class ShortestPathBinaryMatrix {

    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (grid[0][0] != 0 || grid[n - 1][n - 1] != 0) {
            return -1;
        }

        int[][] pos = new int[][]{
            {-1, -1}, {-1, 0}, {-1, 1},
            {0, -1},           {0, 1},
            {1, -1},  {1, 0},  {1, 1}
        };


        Queue<int[]> queue = new ArrayDeque();
        queue.add(new int[] {0, 0});
        int count = 1;
        while(!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] current = queue.poll();
                int x = current[0];
                int y = current[1];
                if (x < 0 || y < 0 || x >= n || y>= n) {
                    continue;
                }
                if (x == n - 1 && y == n - 1) {
                    return count;
                }
                if (grid[x][y] == 1) {
                    continue;
                }
                grid[x][y] = 1;
                for (int[] p: pos) {
                    queue.add(new int[]{current[0] + p[0], current[1] + p[1]});
                }
            }
            count++;
        }
        return -1;
    }
}
